lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,390 @@
+				foreach ( $selector as $s ) {
+					if ( 'element' === $s['type'] && $tags->get_tag() === strtoupper( $s['identifier'] ) ) {
+						return $tags;
+					}
+
+					// @TODO: $tags->has_class() would be _really_ handy here.
+					if ( 'class' === $s['type'] && preg_match( "~\b{$s['identifier']}\b~", $tags->get_attribute( 'class' ) ) ) {
+						return $tags;
+					}
+
+					if ( 'hash' === $s['type'] && $s['identifier'] === $tags->get_attribute( 'id' ) ) {
+						return $tags;
+					}
+				}
+			}
+
+			return null;
+		}
+
+		switch ( $selector['type'] ) {
+			case 'element':
+				$tags->next_tag( [ 'tag_name' => $selector['identifier'] ] );
+				return $tags;
+
+			case 'class':
+				$tags->next_tag( [ 'class_name' => $selector['identifier'] ] );
+				return $tags;
+
+			case 'hash':
+				while ( $tags->next_tag() ) {
+					if ( $selector['identifier'] === $tags->get_attribute( 'id' ) ) {
+						return $tags;
+					}
+				}
+		}
+
+		return null;
+	}
+
+	public static function get_inner_html( WP_HTML_Naive_Processor $tags ) {
+		$tags->set_bookmark( 'start' );
+		$tag_name = $tags->get_tag();
+		$depth = 1;
+
+		if ( self::is_void_element( $tag_name ) ) {
+			return '';
+		}
+
+		while ( $tags->next_tag( [ 'tag_closers' => 'visit' ] ) ) {
+			if ( $tags->get_tag() !== $tag_name ) {
+				continue;
+			}
+
+			if ( $tags->is_tag_closer() && $depth === 1 ) {
+				$tags->set_bookmark( 'end' );
+				break;
+			}
+
+			$depth += $tags->is_tag_closer() ? -1 : 1;
+		}
+
+		return $tags->inner_content( 'start', 'end' );
+	}
+
+	/**
+	 * @see https://html.spec.whatwg.org/#elements-2
+	 */
+	public static function is_void_element( $tag_name ) {
+		switch ( $tag_name ) {
+			case 'area':
+			case 'base':
+			case 'br':
+			case 'col':
+			case 'embed':
+			case 'hr':
+			case 'img':
+			case 'input':
+			case 'link':
+			case 'meta':
+			case 'source':
+			case 'track':
+			case 'wbr':
+				return true;
+
+			default:
+				return false;
+		}
+	}
+
+	public static function parse_definition( $definition ) {
+		if ( empty( $definition['source'] ) ) {
+			return 'not-sourced';
+		}
+
+		$source = $definition['source'];
+		if ( 'html' !== $source && 'attribute' !== $source ) {
+			return 'unsupported';
+		}
+
+		if ( 'attribute' === $source && empty( $definition['selector'] ) ) {
+			return null;
+		}
+
+		if ( 'html' === $source && empty( $definition['selector'] ) ) {
+			return 'inner-html';
+		}
+
+		$selector = self::parse_selector( $definition['selector'] );
+		if ( null === $selector ) {
+			return 'unsupported';
+		}
+
+		if ( 'html' === $source ) {
+			return array( 'type' => 'html', 'selector' => $selector );
+		}
+
+		$attribute = self::parse_attribute( $definition['attribute'] );
+		if ( null === $attribute ) {
+			return null;
+		}
+
+		return array( 'type' => 'attribute', 'selector' => $selector, 'attribute' => $attribute );
+	}
+
+	public static function parse_selector( $s, $at = 0 ) {
+		$selectors = explode( ',', $s );
```

Comment:
I'd treat the separator more carefully in case this gets applied to a selector like `a[href*=","]`

Reply:
Also, the explode will be executed every time this function runs, including the call in line 254 below where we already expect to deal with a single selector. only.

Reply:
yep, thanks. still very rough, and in fact I want to move a lot of code around here, but I wanted to get some basic structure together to see how to build this

Reply:
I've moved support for this into the main parser function and am aborting on unsupported attribute queries, so we _shouldn't_ have this bug anymore